protocol/l2-forked-tests.md

```diff
@@ -0,0 +1,209 @@
+        ///            calling `L2ProxyAdmin.upgrade(address(L1BlockProxy), address(L1BlockImpl))`.
+        ///         3. Call `L1Block.setXFork()` to pull the values from L2 contracts.
+        ///         4. Upgrades the remainder of the L2 contracts via `L2ProxyAdmin.upgrade()`.
+    }
+}
+
+```
+
+1. **Execution Script**: A specialized module that manages the sequential execution of defined NUTs
+
+```solidity
+contract Upgrader {
+    function upgrade(NUTExecutor[] memory _executors) external {
+	    // Apply NUTExecutors based on the initial state
+    }
+}
+
+```
+
+This approach facilitates the specification of NUT sets to be executed from a defined initial state. The implementation would be encapsulated within a contract similar to ForkLive and serve as an alternative to the L2Genesis script, to be invoked during the Setup phase.
+
+### Integration with Client Upgrade Process
+
+Currently, we use Go scripts to prepare NUTs for each upgrade, which are then passed back to the pipeline. These NUTs are defined in dedicated Go files that build them individually. Ideally, we should share these transaction definitions between the client and our test suite to eliminate code duplication and ensure that our test upgrades closely mirror production environments.
+
+Once we have contracts that handle entire sets of NUTs, the transactions in the Go scripts can be replaced by two standard transactions:
+
+1. Deploying the appropriate NUTExecutor contract for that particular upgrade
+2. Calling the execute function on that contract
+
+This approach offers flexibility and power as the exact same transactions could be executed both in production upgrades and our Solidity test suite (provided we use the same arguments). Additionally, it would enable the test suite to easily fuzz transactions as needed.
+
+**Advantages:**
+
+- Provides high-fidelity representation of Predeploy states by simulating complete upgrade paths from an initial network state
+- Offers flexible configuration options for various testing scenarios
+
+However, current limitations prevent us from implementing a truly shared approach, as we require NUT scripts to execute within the context of privileged accounts such as ProxyAdmin or the Depositor Account. To address this, we propose the implementation of the following solution:
+
+#### L2 `ProxyAdmin` Upgrade
+
+It's possible to upgrade the L2 `ProxyAdmin` contract to allow limited delegated calls, enabling NUT scripts to execute within the `ProxyAdmin` context and perform contract upgrades. While this change would allow us to deploy new implementations and upgrade the proxies, configuration settings in these new implementations can be managed by verifying `tx.origin` rather than `msg.sender` as the Depositor Account. We consider this approach semantically correct as it still verifies that the Depositor Account is the originator of the transaction.
```

Comment:
Hmmm, OK, I think I see what this is trying to solve. 

With L1 upgrades, the ProxyAdmin owner is a Safe, so we can just delegate call the OPCM and have the upgrade logic run in the context of the ProxyAdmin owner, which results in most calls being made through the ProxyAdmin.

But on L2, NUTs have typically impersonated the zero address so that they can make upgrade calls directly to the Proxy contracts, and we don't have any way of making a delegatecall from the ProxyAdmin owner to the ProxyAdmin (do we?). It would be nice if we could have the NUTExecutor look and feel a lot like the OPCM, and maybe even reuse a lot of the underlying logic.